misc/openvas-krb5.c

```diff
@@ -0,0 +1,644 @@
+          free (context->gss_channel_bindings);
+        }
+      if (context->gss_actual_mech_type != NULL)
+        {
+          gss_release_oid (NULL, &context->gss_actual_mech_type);
+        }
+      free (context);
+    }
+}
+
+OKrb5ErrorCode
+o_krb5_gss_prepare_context (const OKrb5Credential *creds,
+                            struct OKrb5GSSContext *gss_context)
+{
+  char *target_principal_str = NULL;
+  OKrb5ErrorCode result = O_KRB5_SUCCESS;
+
+  gss_name_t gss_target = GSS_C_NO_NAME;
+  OM_uint32 maj_stat;
+  OM_uint32 min_stat;
+  gss_buffer_desc targetbuf = GSS_C_EMPTY_BUFFER;
+  const struct OKrb5Target *target = &creds->target;
+
+  if (gss_context->gss_creds == GSS_C_NO_CREDENTIAL)
+    {
+      if ((result = okrb5_gss_authenticate (creds, gss_context)))
+        {
+          goto result;
+        }
+    }
+
+  if (target->domain.len != 0)
+    {
+      ALLOCATE_AND_CHECK (target_principal_str, char,
+                          target->host_name.len + target->domain.len
+                            + target->service.len + creds->realm.len + 4,
+                          result);
+      sprintf (target_principal_str, "%s/%s/%s@%s",
+               (char *) target->service.data, (char *) target->host_name.data,
+               (char *) target->domain.data, (char *) creds->realm.data);
+    }
+  else
+    {
+      ALLOCATE_AND_CHECK (target_principal_str, char,
+                          target->host_name.len + target->service.len
+                            + creds->realm.len + 3,
+                          result);
+      sprintf (target_principal_str, "%s/%s@%s", (char *) target->service.data,
+               (char *) target->host_name.data, (char *) creds->realm.data);
+    }
+
+  targetbuf = (gss_buffer_desc){
+    .value = target_principal_str,
+    .length = strlen (target_principal_str),
+  };
+
+  maj_stat = gss_import_name (&min_stat, &targetbuf,
+                              // might also be GSS_C_NT_HOSTBASED_SERVICE,
+                              // but samba uses GSS_C_NT_USER_NAME
+                              GSS_C_NT_USER_NAME, &gss_target);
+  if (maj_stat != GSS_S_COMPLETE)
+    {
+      result = O_KRB5_ERROR + maj_stat;
+      goto result;
+    }
+
+  gss_context->gss_target = gss_target;
+  // gss_set_neg_mechs() already specified that we want gss_mech_krb5
+  // and/or gss_mech_iakerb
+  // so we use spnego to do the negotiation
+  gss_context->gss_mech = gss_mech_spnego;
+  gss_context->gss_want_flags = GSS_C_MUTUAL_FLAG | GSS_C_DELEG_POLICY_FLAG
+                                | GSS_C_REPLAY_FLAG | GSS_C_SEQUENCE_FLAG
+                                | GSS_C_INTEG_FLAG | GSS_C_CONF_FLAG;
+  gss_context->gss_got_flags = 0;
+  gss_context->gss_channel_bindings = GSS_C_NO_CHANNEL_BINDINGS;
+  gss_context->gss_time_req = 0;
+  gss_context->gss_time_rec = 0;
+  gss_context->gss_actual_mech_type = NULL;
+result:
+  // TODO: cleanup target_principal_str on failure?
```

Comment:
target_principal_srt is being point to by target_buf, target_buf is being transfered to gss_import_name, therefore it may be used within the gss framework. Before I remove the comment I will double check if they create a clone of that. I doubt it though.

Reply:
nevermind I should read my own comment, yes, on failure I should